java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -90,169 +89,173 @@ private FlowComputation(@Nullable SymbolicValue symbolicValue, Predicate<Constra
-      return methodInvocationFlow(constraint, edge);
+    ExecutionPath(@Nullable ExplodedGraph.Edge edge, PSet<ExplodedGraph.Edge> visited, @Nullable Symbol trackSymbol, List<JavaFileScannerContext.Location> flow, boolean finished) {
+      this.trackSymbol = trackSymbol;
+      this.lastEdge = edge;
+      this.visited = visited;
+      this.flow = flow;
+      this.finished = finished;
     }
-    if (nodeTree.is(Tree.Kind.NEW_CLASS)) {
-      return Stream.of(location(parent, String.format("Constructor implies '%s'.", constraint.valueAsString())));
+
+    ExecutionPath(@Nullable Symbol trackSymbol) {
+      this(null, PCollections.emptySet(), trackSymbol, ImmutableList.of(), false);
     }
-    String name = SyntaxTreeNameFinder.getName(nodeTree);
-    String message = name == null ? constraint.valueAsString() : String.format("Implies '%s' is %s.", name, constraint.valueAsString());
-    return Stream.of(location(parent, message));
-  }
 
-  private Stream<JavaFileScannerContext.Location> methodInvocationFlow(Constraint learnedConstraint, ExplodedGraph.Edge edge) {
-    ExplodedGraph.Node parent = edge.parent;
-    MethodInvocationTree mit = (MethodInvocationTree) parent.programPoint.syntaxTree();
-    Stream.Builder<JavaFileScannerContext.Location> flowBuilder = Stream.builder();
-    SymbolicValue returnSV = edge.child.programState.peekValue();
-    if (symbolicValues.contains(returnSV)) {
-      flowBuilder.add(location(parent, String.format("'%s()' returns %s.", mit.symbol().name(), learnedConstraint.valueAsString())));
+    ExecutionPath addEdge(ExplodedGraph.Edge edge) {
+      ImmutableList.Builder<JavaFileScannerContext.Location> flowBuilder = ImmutableList.builder();
+      flowBuilder.addAll(flow);
+      Symbol newTrackSymbol = addFlowFromLearnedAssociations(edge, flowBuilder);
+      Stream<Constraint> constraintStream = addFlowFromLearnedConstraints(edge, flowBuilder);
+      return new ExecutionPath(edge, visited.add(edge), newTrackSymbol, flowBuilder.build(), edge.parent.edges().isEmpty() || constraintStream.anyMatch(terminateTraversal));
     }
-    SymbolicValue invocationTarget = parent.programState.peekValue(mit.arguments().size());
-    if (symbolicValues.contains(invocationTarget)) {
-      flowBuilder.add(location(parent, "..."));
+
+    boolean notVisited(ExplodedGraph.Edge e) {
+      return !visited.contains(e);
     }
-    int argIdx = correspondingArgumentIndex(symbolicValues, parent);
-    if (argIdx != -1) {
-      ExpressionTree argTree = mit.arguments().get(argIdx);
-      String message = String.format("Implies '%s' is %s.", SyntaxTreeNameFinder.getName(argTree), learnedConstraint.valueAsString());
-      flowBuilder.add(new JavaFileScannerContext.Location(message, argTree));
+
+    @Nullable
+    private Symbol addFlowFromLearnedAssociations(ExplodedGraph.Edge edge, ImmutableList.Builder<JavaFileScannerContext.Location> flowBuilder) {
+      Symbol newTrackSymbol = trackSymbol;
```

Comment:
newTrackSymbol is a bit useless here : either remove it (personal pref) or assign it in the isPresent if (to get rid of the direct return).